tools/generate_release_notes.py

```diff
@@ -1,246 +1,420 @@
+                logger.info(
+                    "pull request associated with multiple commits: %r",
+                    pull.html_url,
+                )
+        all_pull_requests.update(commit_pull_requests)
+    return all_pull_requests
+
+
+@dataclass(frozen=True, eq=True)
+class UnknownCoAuthor:
+    """Represents a co-author for which the GitHub user is not known.
+
+    Hashing and comparing only takes into account the `name` attribute.
+    """
+
+    name: str
+    email: str = field(compare=False)
+    commit: Commit = field(compare=False)
 
 
-all_commits = list(
-    tqdm(
-        repository.get_commits(sha=args.to_commit, since=previous_tag_date),
-        desc=f"Getting all commits between {args.from_commit} " f"and {args.to_commit}",
+def _find_coauthors(commit: Commit) -> set[UnknownCoAuthor]:
+    co_author_regex = re.compile(
+        r"^\s*Co-authored-by: (?P<name>[^<]+) <(?P<email>[^>]+)>$", flags=re.MULTILINE
     )
-)
-all_hashes = {c.sha for c in all_commits}
+    message = commit.commit.message
+    matches = co_author_regex.finditer(message)
+    matches = list(matches)
+    coauthors = {
+        UnknownCoAuthor(name=m["name"], email=m["email"], commit=commit)
+        for m in matches
+    }
+    return coauthors
+
+
+def contributors(
+    commits: Iterable[Commit], pull_requests
+) -> tuple[set[NamedUser], set[Union[NamedUser, UnknownCoAuthor]], set[NamedUser]]:
+    """Fetch code authors and reviewers.
+
+    `authors` are first authors of commits; co-authors are not included (yet).
+    `reviewers` are users, who added reviews to a merged pull request or create the
+    merge commit for one.
+    """
+    authors = set()
+    reviewers = set()
+    unknown_coauthors = set()
+
+    for commit in commits:
+        if commit.author:
+            authors.add(commit.author)
+        if commit.committer:
+            reviewers.add(commit.committer)
+        unknown_coauthors.update(_find_coauthors(commit))
+
+    for pull in pull_requests:
+        for review in pull.get_reviews():
+            if review.user:
+                reviewers.add(review.user)
+
+    # Try to replace unknown coauthors with known users by matching email
+    coauthors = set()
+    known_users_by_mail = {u.email: u for u in authors | reviewers}
+    for unknown in unknown_coauthors:
+        if unknown.email in known_users_by_mail:
+            coauthors.add(known_users_by_mail[unknown.email])
+        else:
+            coauthors.add(unknown)
 
-authors = set()
-reviewers = set()
-committers = set()
-users = dict()  # keep track of known usernames
+    return authors, coauthors, reviewers
 
 
-def find_author_info(commit):
-    """Return committer and author of a commit.
+@dataclass(frozen=True, kw_only=True)
+class MdFormatter:
+    """Format release notes in Markdown from PRs, authors and reviewers."""
 
-    Parameters
-    ----------
-    commit : Github commit
-        The commit to query.
+    pull_requests: set[PullRequest]
```

Comment:
I don't know this docstring format?

Reply:
Huh? This is not a docstring. It's a type hint to a dataclass...